ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLAgentExecutor.java

```diff
@@ -113,41 +122,51 @@ public void execute(Input input, ActionListener<Output> listener) {
                             MLMemorySpec memorySpec = mlAgent.getMemory();
                             String memoryId = inputDataSet.getParameters().get(MEMORY_ID);
                             String parentInteractionId = inputDataSet.getParameters().get(PARENT_INTERACTION_ID);
+                            String regenerateInteractionId = inputDataSet.getParameters().get(REGENERATE_INTERACTION_ID);
                             String appType = mlAgent.getAppType();
                             String question = inputDataSet.getParameters().get(QUESTION);
 
+                            if (memoryId == null && regenerateInteractionId != null) {
+                                listener.onFailure(new MLValidationException("Memory id must provide for regenerate"));
+                            }
+
                             if (memorySpec != null
                                 && memorySpec.getType() != null
                                 && memoryFactoryMap.containsKey(memorySpec.getType())
                                 && (memoryId == null || parentInteractionId == null)) {
                                 ConversationIndexMemory.Factory conversationIndexMemoryFactory =
                                     (ConversationIndexMemory.Factory) memoryFactoryMap.get(memorySpec.getType());
-                                conversationIndexMemoryFactory.create(question, memoryId, appType, ActionListener.wrap(memory -> {
+                                conversationIndexMemoryFactory.create(question, memoryId, appType, wrap(memory -> {
                                     inputDataSet.getParameters().put(MEMORY_ID, memory.getConversationId());
-                                    ConversationIndexMemory conversationIndexMemory = (ConversationIndexMemory) memory;
+
+                                    // get regenerate interaction question
+                                    Optional.ofNullable(regenerateInteractionId).ifPresent(interactionId -> {
+                                        log.info("Regenerate for existing interaction {}", regenerateInteractionId);
+                                        getQuestionFromInteraction(memoryId, interactionId, inputDataSet);
+                                    });
+
                                     // Create root interaction ID
                                     ConversationIndexMessage msg = ConversationIndexMessage
                                         .conversationIndexMessageBuilder()
                                         .type(appType)
-                                        .question(question)
+                                        .question(inputDataSet.getParameters().get(QUESTION))
```

Comment:
Why not use question?

Reply:
question will be override at line 204 for regenerate case

Reply:
Could you add comment for this specific regenerate case? It takes a while to figure out why we can't use the question in request directly

Reply:
thanks, comments added.

Reply:
Yeah, I was confused too. I think you can also set question to the original question meanwhile updating inputDataSet.

Reply:
refactor to listener way and it's more clear now.